### PR TITLE
Self serve migration

### PR DESCRIPTION
**Pull request by RensR**
This PR will add two new pools, one burnMint and one lockRelease. These will be used as proxies for the existing <1.5 token pools when the lanes upgrade to 1.5. Please see `contracts/src/v0.8/ccip/test/legacy/BurnMintTokenPoolAndProxy.t.sol` for a full step-by-step guide on how this upgrade works for both the 1.0/1.2 and the 1.4 pools. 

Pools deployed after self serve is live will NOT use these proxies, but rather the clean versions that do not implement the proxy behaviour. Because of this, we only need proxies for the pools that are currently live: burnMint and LockRelease. USDC will cleanly upgrade by deploying new pools and doesn't require the proxy. All other pool flavours are not live at the moment

### CODE REVIEW

File: contracts/src/v0.8/ccip/pools/LockReleaseTokenPoolAndProxy.sol [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+import {TokenPool} from "./TokenPool.sol";
+
+import {IERC20} from "../../vendor/openzeppelin-solidity/v4.8.3/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20} from "../../vendor/openzeppelin-solidity/v4.8.3/contracts/token/ERC20/utils/SafeERC20.sol";
+
+/// @notice Token pool used for tokens on their native chain. This uses a lock and release mechanism.
+/// Because of lock/unlock requiring liquidity, this pool contract also has function to add and remove
+/// liquidity. This allows for proper bookkeeping for both user and liquidity provider balances.
+/// @dev One token per LockReleaseTokenPool.
+contract LockReleaseTokenPoolAndProxy is LegacyPoolWrapper, ILiquidityContainer, ITypeAndVersion {
```
Comment (RensR) [inline contracts/src/v0.8/ccip/pools/LockReleaseTokenPoolAndProxy.sol:19]: Sadly, this pool contract had to duplicate a bit more, as the LockRelease pool has more logic than BurnMint

File: contracts/src/v0.8/ccip/test/legacy/BurnMintTokenPool1_4.sol [hunk 2]
```diff
@@ -0,0 +1,402 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity 0.8.19;
```
Comment (RensR) [inline contracts/src/v0.8/ccip/test/legacy/BurnMintTokenPool1_4.sol:2]: Single-file copy of our 1.4 pool

File: contracts/src/v0.8/ccip/test/legacy/BurnMintTokenPool1_2.sol [hunk 3]
```diff
@@ -0,0 +1,354 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.0;
```
Comment (RensR) [inline contracts/src/v0.8/ccip/test/legacy/BurnMintTokenPool1_2.sol:2]: Single-file copy of our 1.2 pool

**Review by justinkaseman — APPROVED** [review]

File: contracts/src/v0.8/ccip/pools/BurnMintTokenPoolAndProxy.sol [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    returns (Pool.LockOrBurnOutV1 memory)
+  {
+    _checkAllowList(lockOrBurnIn.originalSender);
+    _onlyOnRamp(lockOrBurnIn.remoteChainSelector);
+    _consumeOutboundRateLimit(lockOrBurnIn.remoteChainSelector, lockOrBurnIn.amount);
+
+    if (!_hasLegacyPool()) {
+      IBurnMintERC20(address(i_token)).burn(lockOrBurnIn.amount);
+    } else {
+      _lockOrBurnLegacy(lockOrBurnIn);
```
Comment (matYang) [inline contracts/src/v0.8/ccip/pools/BurnMintTokenPoolAndProxy.sol:37]: 1.2 interface supports return data from `lockOrBurn` to be passed to dest right: https://github.com/smartcontractkit/ccip/blob/contracts-ccip/v1.2.1/contracts/src/v0.8/ccip/onRamp/EVM2EVMOnRamp.sol#L347

likely not used by any customers, would it still be better to pass that along to ensure compatibility.

File: contracts/src/v0.8/ccip/pools/BurnMintTokenPoolAndProxy.sol [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    override
+    whenHealthy
+    returns (Pool.LockOrBurnOutV1 memory)
+  {
+    _checkAllowList(lockOrBurnIn.originalSender);
+    _onlyOnRamp(lockOrBurnIn.remoteChainSelector);
+    _consumeOutboundRateLimit(lockOrBurnIn.remoteChainSelector, lockOrBurnIn.amount);
+
+    if (!_hasLegacyPool()) {
+      IBurnMintERC20(address(i_token)).burn(lockOrBurnIn.amount);
```
Comment (matYang) [inline contracts/src/v0.8/ccip/pools/BurnMintTokenPoolAndProxy.sol:35]: are we gonna ditch the different burn flavours?

File: contracts/src/v0.8/ccip/pools/BurnMintTokenPoolAndProxy.sol (hunk 4 above)
Comment (RensR) [reply contracts/src/v0.8/ccip/pools/BurnMintTokenPoolAndProxy.sol:37]: There are no custom pools live right now, and this proxy will only be used for current pools. The USDC pool is the only one with a return value, and those pools will get a clean deploy of the 1.5 USDC pool

File: contracts/src/v0.8/ccip/pools/BurnMintTokenPoolAndProxy.sol (hunk 5 above)
Comment (RensR) [reply contracts/src/v0.8/ccip/pools/BurnMintTokenPoolAndProxy.sol:35]: This is just the proxy. The other flavours are not live so we don't need a proxy for them.

**Review by matYang — APPROVED** [review]

**Pull request merged**